cocotb/share/include/gpi.h

```diff
@@ -72,7 +72,36 @@ we have to create a process with the signal on the sensitivity list to imitate a
 # define __attribute__(x)
 #endif
 
-
+/*
+ * Declare the handle types.
+ *
+ * We want these handles to be opaque pointers, since their layout is not
+ * exposed to C. We do this by using incomplete types. The assumption being
+ * made here is that `sizeof(some_cpp_class*) == sizeof(some_c_struct*)`, which
+ * is true on all reasonable platforms.
+ */
+#ifdef __cplusplus
+/* In C++, we use forward-declarations of the types in gpi_priv.h as our
+ * incomplete types, as this avoids the need for any casting in GpiCommon.cpp.
+ */
+class GpiObjHdl;
+class GpiCbHdl;
+class GpiIterator;
+typedef GpiObjHdl *gpi_sim_hdl;
+typedef GpiCbHdl *gpi_cb_hdl;
+typedef GpiIterator *gpi_iterator_hdl;
+#else
+/* In C, we declare some incomplete struct types that we never complete.
+ * The names of these are irrelevant, but for simplicity they match the C++
+ * names.
+ */
+struct GpiObjHdl;
+struct GpiCbHdl;
+struct GpiIterator;
+typedef struct GpiObjHdl *gpi_sim_hdl;
+typedef struct GpiCbHdl *gpi_cb_hdl;
+typedef struct GpiIterator *gpi_iterator_hdl;
+#endif
```

Comment:
Like this, I assume?

```suggestion
#ifdef __cplusplus
    /* In C++, we use forward-declarations of the types in gpi_priv.h as our
     * incomplete types, as this avoids the need for any casting in GpiCommon.cpp.
     */
    class GpiObjHdl;
    class GpiCbHdl;
    class GpiIterator;
    typedef GpiObjHdl *gpi_sim_hdl;
    typedef GpiCbHdl *gpi_cb_hdl;
    typedef GpiIterator *gpi_iterator_hdl;
#else
    /* In C, we declare some incomplete struct types that we never complete.
     * The names of these are irrelevant, but for simplicity they match the C++
     * names.
     */
    struct GpiObjHdl;
    struct GpiCbHdl;
    struct GpiIterator;
    typedef struct GpiObjHdl *gpi_sim_hdl;
    typedef struct GpiCbHdl *gpi_cb_hdl;
    typedef struct GpiIterator *gpi_iterator_hdl;
#endif
```

Will change locally so I can squash the commits correctly.

Reply:
:+1: